### PR TITLE
Enhanced KGE Framework: Added TransH, Transfer Learning, New Tasks, Metrics, and Embedding Freezing

### PR DESCRIPTION
**Pull request by HernandezEduin**
This pull request introduces several major improvements to the Knowledge Graph Embedding (KGE) framework:

---

### ✅ New Model Support
- **TransH**: Implemented the TransH model, including relation-specific hyperplane projection logic.

---

### 💾 Checkpointing Enhancements
- **Best Model Saving**: Automatically saves the best model during training based on a user-defined evaluation metric.
- **Folder Cleanup**: Cleans up checkpoints when no model is saved, preventing folder bloat during hyperparameter sweeps.

---

### 🔁 Transfer Learning & Embedding Freezing
- **Embedding Reload**: Allows reusing pretrained entity/relation embeddings across different datasets.
- **Embedding Freezing**: Optionally freeze entity/relation embeddings during training to enable fine-tuning scenarios.

---

### 🧠 Extended KGE Training Tasks

This PR introduces several new **auxiliary training tasks** to enhance embedding robustness and structural awareness in Knowledge Graph Embedding (KGE) models. These tasks are **compatible with existing architectures** and **share parameters** with the primary link prediction model, enabling seamless multi-task learning.

Each task can be enabled independently and is **weighted during training using a configurable λ (lambda) coefficient**, which controls its relative impact compared to the main link prediction objective. When wildcards (`*`) are used in the task definitions, either a **shared wildcard embedding** is employed or predictions are **averaged across all valid candidates**, based on the chosen configuration.

During evaluation, standard metrics such as **MRR, MR, and Hits@K** are **reported per task** and also **averaged across all tasks**, providing a comprehensive measure of performance known as the **Multi-Task Metric**. This helps assess the model's generalization across diverse prediction scenarios.

#### 🔹 Relation Prediction (h, ?, t)
- **Objective:** Predict the most likely relation between a given head and tail entity.
- **Use Case:** Completes partially specified triples when the relation is missing.
- **Training Input:** (h, ?, t)  
- **Training Output:** A probability distribution over all relation types.

#### 🔹 Domain Prediction ( (?, r, *), (*, r, t) )
- **Objective:** Given one entity and a relation, predict the missing entity from the appropriate domain (head or tail).
- **Use Case:** Enforces type consistency and relation semantics during training.
- **Training Input:** (?, r, *), (*, r, t)  
- **Training Output:** Candidate head or tail entity scores for the specified relation.

#### 🔹 Neighborhood Entity Prediction (NEP): (h, *, ?), (?, *, t)
- **Objective:** Predict plausible entities that neighbor a given entity, without explicit relation knowledge.
- **Use Case:** Teaches structural proximity and implicit neighborhood understanding.
- **Training Input:** Entity with partial triple (h, *, ?) or (?, *, t)  
- **Training Output:** Likely tail or head entities based on shared structural patterns.

#### 🔹 Neighborhood Relation Prediction (NRP): (h, ?, *), (*, ?, t)
- **Objective:** Predict the most likely relation linking a given entity to its neighbors.
- **Use Case:** Enhances the model's ability to infer edge types from local structure.
- **Training Input:** Entity and wildcard (h, ?, *), (*, ?, t)  
- **Training Output:** Likely relation(s) between the given entity and its neighborhood.
---

### 📊 New Evaluation Metrics

- **Semantic Recall**  
  Evaluates how often the top-*k* predicted entities belong to the correct semantic category or type (e.g., *Person*, *Location*). This metric provides **type-aware feedback** on prediction quality.  
  - Scores range from **0 to 1**, with **higher values indicating better performance**.
  - The metric becomes **stricter as *k* increases**.
  - If the number of valid types *n* is smaller than *k*, evaluation defaults to **top-*n***.

- **NBE-Recall** *(Neighborhood-Based Entity Recall)*  
  Measures the fraction of top-*k* predicted entities that are **true neighbors of the query node** in the knowledge graph. This metric reflects how well the model captures **local structural accuracy** around entities.

- **NBR-Recall** *(Neighborhood-Based Relation Recall)*  
  Evaluates how many of the top-*k* predicted relations correctly **connect the query entity to its known neighbors**. It reflects the model's ability to preserve **local relational consistency** in the graph structure.

### CODE REVIEW

**Review by ottersome — DISMISSED** [review]
Just *amend* this commit delete the model tbh. Its not being used anywhere for four months now. 
Next time if you commit an entire chunk please make sure you commit spaces in betwen, it helps with knowing when the comment starts and ends, also helps with LSP tool :] !.
image

**Comment by HernandezEduin** [comment]
> Just *amend* this commit delete the model tbh. Its not being used anywhere for four months now. 
> Next time if you commit an entire chunk please make sure you commit spaces in betwen, it helps with knowing when the comment starts and ends, also helps with LSP tool :] !.
> image

Could you perhaps be more specific? Which commit? Which model? What do you mean by "commit chunks"?

**Comment by ottersome** [comment]
> > Just _amend_ this commit delete the model tbh. Its not being used anywhere for four months now.
> > Next time if you commit an entire chunk please make sure you commit spaces in betwen, it helps with knowing when the comment starts and ends, also helps with LSP tool :] !.
> > image
> 
> Could you perhaps be more specific? Which commit? Which model? What do you mean by "commit chunks"?

PMing you with deets

File: multihopkg/exogenous/sun_models.py [hunk 1]
```diff
@@ -996,6 +1000,76 @@ def calculate_entity_range(embeddings: Union[nn.Embedding, nn.Parameter]):
...
+def save_configs(args):
+    '''
+    Save the configurations to a json file
+    '''
+    
+    argparse_dict = vars(args)
+    with open(os.path.join(args.save_path, 'config.json'), 'w') as fjson:
+        json.dump(argparse_dict, fjson)
+
+def save_model(model, optimizer, save_variable_list, save_dir, autoencoder_flag=False):
```
Comment (ottersome) [inline multihopkg/exogenous/sun_models.py:1481]: Can we move these functions (`save_configs`, `save_model`,..etc) instead to another file ? 
Perhaps `<repo_root>/multihopkg/utils/ops.py`, or just a file that is for operations on models, your choice. 
`exogeneous` directory is meant for algorithms/ideas that originate _externally_, such as Sun et. al models. 
And along the way, please also move: `get_embeddings_from_indices`,`calculate_entity_centroid`,`calculate_entity_range`,
into the same file since these functions are all very general and not limited to Sun et al model 😅.

File: multihopkg/exogenous/sun_models.py [hunk 2]
```diff
@@ -996,422 +996,422 @@ def calculate_entity_range(embeddings: Union[nn.Embedding, nn.Parameter]):
         min_range = torch.min(embeddings.weight.data).item()
     return min_range, max_range
 
-class LegacyKGEModel(nn.Module):
```
Comment (ottersome) [inline multihopkg/exogenous/sun_models.py:999]: You can delete *this* model rather than commenting it. Its been on Legacy state for a while now.

Comment (ottersome) [reply multihopkg/exogenous/sun_models.py:999]: (Just ammend the commit *e4ef0e1ee05c68ad4b69462716fe72ea6aa8d3f6*. No real need to create a new one )

**Comment by ottersome** [comment]
> > Just _amend_ this commit delete the model tbh. Its not being used anywhere for four months now.
> > Next time if you commit an entire chunk please make sure you commit spaces in betwen, it helps with knowing when the comment starts and ends, also helps with LSP tool :] !.
> > image
> 
> Could you perhaps be more specific? Which commit? Which model? What do you mean by "commit chunks"?

The commit ~chunk~ hunk thing (sorry for the previous typo) is mostly a styling issue that imo leads to poor readability on the long term. If you are interested in more detail I'm happy to elaborate. Otherwise the explanation is boring and meaningless. Just do you.

As for the other specifics about commit and mode, I made new revisions that are more clear.

**Comment by ottersome** [comment]
The rest looks good 👍

**Comment by HernandezEduin** [comment]
You mean to remove the Legacy Model from the commits?

**Comment by HernandezEduin** [comment]
Well, that was painful.
@ottersome How about now?

File: multihopkg/exogenous/sun_models.py [hunk 3]
```diff
@@ -590,44 +796,288 @@ def test_step(model, test_triples, all_true_triples, args):
...
+                                )
+
+                                log_entry |= KGEModel.get_recall(
+                                    argsort_raw[i, :], constraint_set, k_values, 'NBE'
+                                )
+
+                            # NBR@K evaluation using raw metrics (only for relation/nbr-head/nbr-tail prediction)
+                            if nbr_k_condition:
+                                if mode == "relation-batch":
+                                    constraint_set = constraints["head_neighborhood_rel_constraints"].get(positive_arg[i].item(), set()) or constraints["tail_neighborhood_rel_constraints"].get(positive_arg[i].item(), set())
```
[the remark this answers is not in the document]
Comment (HernandezEduin) [reply multihopkg/exogenous/sun_models.py]: I'm combining sets, so it should be `|` instead of `or`

[the remark this answers is not in the document]
Comment (HernandezEduin) [reply multihopkg/exogenous/sun_models.py]: Corrected in 7aeeffc

**Comment by ottersome** [comment]
Pay no attention to the above. I needed to do it before merging the PR.

**Comment by ottersome** [comment]
I would still like to move the `save_model` and similar general functions withing `sun_models` elsewhere. But I'll do that myself in the next PR. For now, thanks. 辛苦了

**Review by ottersome — APPROVED** [review]
